### DISCUSSION TITLE
Using disposables around sideeffects.

### DISCUSSION THREAD

**Discussion by adameasten** [louthy/language-ext#1161] [Q&A] [answered]
Hello! I’m new to functional programming and having troubles when lazy evaluating a bunch of sideeffects combined with opentelemtry and measuring calls to a database etc. The lazy evaluation of my sideeffects of type Eff executes with a single run att the end of the declaration of all expressions. This results in all usings only covers the lazy init of each sideeffect, not the actual execution! Anyone have faced the same problem?

**Comment by louthy** [comment, marked as answer]
If you're using disposables that have been created outside of an `Eff` or `Aff` then you will need to force the evaluation of the `Eff` or `Aff` before disposing.  Otherwise it's possible to _lift_ disposables into the `Eff` or `Aff` expression, so they're run lazily too:

```c#
    var openFile = Eff<FileStream>(() => File.OpenText(path));

    use(openFile, stream => /* Eff expression that uses the stream */);
```

**Reply by adameasten** [reply]
Thank you! I missed the existence of use(..)! I'll try it out :)
